src/devices/bus/isa/sblaster_lle.cpp

```diff
@@ -0,0 +1,381 @@
+	m_ym3812(*this, "ym3812"),
+	m_dac(*this, "dac"),
+	m_speaker(*this, "speaker")
+{
+}
+
+void isa8_sblaster_2_0_lle_device::device_start()
+{
+	set_isa_device();
+}
+
+void isa8_sblaster_2_0_lle_device::device_reset()
+{
+	uint16_t base_port = (ioport("JP1")->read() & 0x02) ? 0x240 : 0x220;
+
+	m_isa->install_device(0x0200, 0x0207,
+		read8smo_delegate(*subdevice<pc_joy_device>("pc_joy"), FUNC(pc_joy_device::joy_port_r)),
+		write8smo_delegate(*subdevice<pc_joy_device>("pc_joy"), FUNC(pc_joy_device::joy_port_w)));
+
+	m_isa->install_device(base_port + 0x6, base_port + 0x7,
+		read8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_reset_r)),
+		write8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_reset_w)));
+	m_isa->install_device(base_port + 0x8, base_port + 0x9,
+		read8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::ym3812_16_r)),
+		write8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::ym3812_16_w)));
+	m_isa->install_device(base_port + 0xa, base_port + 0xb,
+		read8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_data_r)),
+		write8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_data_w)));
+	m_isa->install_device(base_port + 0xc, base_port + 0xd,
+		read8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_wbuf_status_r)),
+		write8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_cmd_w)));
+	m_isa->install_device(base_port + 0xe, base_port + 0xf,
+		read8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_rbuf_status_r)),
+		write8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::dsp_rbuf_status_w)));
+
+	m_isa->install_device(0x0388, 0x0389,
+		read8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::ym3812_16_r)),
+		write8sm_delegate(*this, FUNC(isa8_sblaster_2_0_lle_device::ym3812_16_w)));
+
+	m_isa->set_dma_channel(1, this, false);
+
+	m_irq_in_flag = false;
+	m_irequest = true;
+	m_drequest = false;
+
+	dsp_reset_w(0, 0);
+
+	m_irq_raised = false;
+	m_dma_raised = false;
+}
+
+// ISA DMA handling
+uint8_t isa8_sblaster_2_0_lle_device::dack_r(int line)
+{
+	// Shouldn't happen.
+	return dsp_data_r(0);
+}
+
+void isa8_sblaster_2_0_lle_device::dack_w(int line, uint8_t data)
+{
+	// Incoming DMA data!
+	lower_dma();            // Acknowledge clears the DMA request.
+	dsp_cmd_w(0, data);     // Write the byte to the data latch.
+}
+
+void isa8_sblaster_2_0_lle_device::raise_dma()
+{
+	// Not configurable.
+	LOG("SB raising DMA %d\n", 1);
+	m_isa->drq1_w(ASSERT_LINE);
+}
+
+void isa8_sblaster_2_0_lle_device::lower_dma()
+{
+	// Not configurable.
+	LOG("SB lowering DMA %d\n", 1);
+	m_isa->drq1_w(CLEAR_LINE);
+}
+
+void isa8_sblaster_2_0_lle_device::raise_irq()
+{
+	if(!m_irq_raised)
+	{
+		LOG("SB raising IRQ %d\n", 5);
+		switch(ioport("JP4")->read())
+		{
+			case 1: m_isa->irq7_w(ASSERT_LINE); break;
+			case 2: m_isa->irq5_w(ASSERT_LINE); break;
+			case 4: m_isa->irq3_w(ASSERT_LINE); break;
+			case 8: m_isa->irq2_w(ASSERT_LINE); break;
+		}
```

Comment:
Don’t do tag lookups on hot paths.  Use object finders.  Also, you should probably keep the interrupt number stored somewhere, otherwise someone could flip the jumper while it’s asserted, and then you’ll deassert the wrong IRQ.